drivers/mcp2515/candev_mcp2515.c

```diff
@@ -0,0 +1,726 @@
+/*
+ * Copyright (C) 2016 OTA keys S.A.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_mcp2515
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the CAN controller driver
+ *
+ * @author      Toon Stegen <toon.stegen@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ * @}
+ */
+
+#include <errno.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "candev_mcp2515.h"
+#include "can/common.h"
+#include "can/device.h"
+#include "mcp2515.h"
+#include "mutex.h"
+#include "periph_conf.h"
+#include "thread.h"
+#include "sched.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+mutex_t mcp_mutex;
+static int neednewisr = 0;
+
+static int _init(candev_t *candev);
+static int _send(candev_t *candev, const struct can_frame *frame);
+static void _isr(candev_t *candev);
+static int _set(candev_t *candev, canopt_t opt, void *value, size_t value_len);
+static int _get(candev_t *candev, canopt_t opt, void *value, size_t max_len);
+static int _abort(candev_t *candev, const struct can_frame *frame);
+static int _set_filter(candev_t *dev, const struct can_filter *filter);
+static int _remove_filter(candev_t *dev, const struct can_filter *filter);
+
+static void _irq_rx(candev_mcp2515_t *dev, int handle);
+static void _irq_tx(candev_mcp2515_t *dev, int handle);
+static void _irq_error(candev_mcp2515_t *dev);
+static void _irq_message_error(candev_mcp2515_t *dev);
+static void _irq_wakeup(const candev_mcp2515_t *dev);
+static void _send_event(const candev_mcp2515_t *dev, candev_event_t event,
+                        void *arg);
+
+static const candev_driver_t candev_mcp2515_driver = {
+    .send = _send,
+    .init = _init,
+    .isr = _isr,
+    .get = _get,
+    .set = _set,
+    .abort = _abort,
+    .set_filter = _set_filter,
+    .remove_filter = _remove_filter,
+};
+
+static const struct can_bittiming_const bittiming_const = {
+    .tseg1_min = 3,     /**< Time segment 1 = prop_seg + phase_seg1, min value */
+    .tseg1_max = 16,    /**< Time segment 1, max value */
+    .tseg2_min = 2,     /**< Time segment 2 = phase_seg2, min value */
+    .tseg2_max = 8,     /**< Time segment 2, max value */
+    .sjw_max = 4,       /**< Synchronisation jump width */
+    .brp_min = 1,       /**< Bit-rate prescaler, min value */
+    .brp_max = 64,      /**< Bit-rate prescaler, max value */
+    .brp_inc = 1,       /**< Bit-rate prescaler, increment */
+};
```

Comment:
I'm unfamiliar with CAN, are these values constant defaults? Are they standard or do they come from the datasheet?

Reply:
I'm no expert on CAN myself, but the way I understand it, these values are related to the bittimings. You can find the formulas for them in the datasheet, or you can find calculators online. They can be changed to tweak the settings of the CAN-bus.

This format is parallel with the other definitions of can_bittimings in RIOT. For example `cpu/stm32/periph/can.c`.

Reply:
Those parameters are what the controller support. They are constant and come from the datasheet. They are used to set up the controller properly from a bitrate and a sampling point provided by a user (see `raw_can_set_bitrate()` for more details).